### DISCUSSION TITLE
Glofas v3.1 data is downloaded, but may have formatting issues

### DISCUSSION THREAD

**Discussion by uwagura** [NOAA-CEFI-Regional-Ocean-Modeling/seasonal-workflow#3] [General]
@yichengt900 , @andrew-c-ross I have successfully downloaded all of the GloFAS v3.1 data from 1979 to September 25th of this year ( the last available day for this dataset ), as well as GloFAS v4.0 data from 2004 to present . A few small issues I've noticed so far:

1. The time variable is named `valid_time`, not `time`
2. There is some floating point imprecision in the lat/lon coordinates, meaning the data sets contain values like `88.8500000000001`

These are easy fixes, if we think they will be an issue. I was going to go ahead and fix these issues using xarray, but I've been having several issues working with the GloFAS data in xarray. I _think_ these issues stem from the formatting of the data, since GloFAS data is natively stored as GRIB files using WMO standards and then converted to netcdf files with the same conventions ( though interestingly, the attributes in the files claim to follow cf conventions). Issues I've had are: 

1. Xarray can't determine the size of the files correctly. It assumes that discharge data is stored as `float32`, so it estimates the size of the datasets as `dimensions * 4 bytes = ~8GB`, when in reality the v3.1 files are only about 800 megabytes big. 
2. As a result of this, it looks like you get weird issues when you try to modify the data in Xarray and then save it as netcdf files. I tried reformatting the metadata to be cf compliant and changing the name of `valid_time` to `time` for the glofas v4.0 data, and when I saved the resulting dataset as a netcdf file it doubled in size for some reason. 
3. In general, opening the dataset, doing computations, and saving it is a slow process, possibly because xarray is inferring a lot of information about the dataset structure. This makes doing any QA/QC on the data a bit more time consuming

None of these are major issues, and I think I could make the relevant changes to coordinate data using nco if need be. Should I try and make these changes, and is it worth looking into the xarray readability issues further? I don't think the underlying data has any issues since I can plot it and it looks ok, though figuring out why xarray has so many issues with the variables may make it easier to work with the data in the future. I should note that GloFAS data that Liz downloaded (`/archive/Liz.Drenkard/datasets/GloFAS/v3.1/netcdf`) has the same xarray issues, so it looks like we can still use the data even if xarray has trouble reading it. 

If we don't think any of these problems are big issues, I now have access to the uda, so I can easily move the data there when needed. Let me know what you think we should do!

P.S: I mentioned this previously, but versions of GloFAS v3.1 downloaded using ecCodes c2.32 or earlier have an incorrect final longitude coordinate. It looks like the version that Liz Download has this problem, have we had any issues working with that data?

**Comment by andrew-c-ross** [comment]
Thanks @uwagura. Could you share the path to the GloFAS v4 data as it is now?

If it doubles in size when you try to re-save the file using xarray, is xarray writing it as float64? You should be able to do something like `.to_netcdf([...], encoding={'dis24': {'dtype': 'float32'}})`

Slight variations in coordinates shouldn't be a big issue since we normally work with one file at a time, but if needed that could be handled when opening the data in xarray with something similar to 
``` python
# Account for very small differences in lat and lon coords by rounding to nearest 1/12 degree.
def round_coords(ds):
    ds['latitude'] = np.round(ds['latitude'] * 12 ) / 12
    ds['longitude'] = np.round(ds['longitude'] * 12) / 12
    return ds

xarray.open_mfdataset([...], preprocess=round_coords)
```

**Reply by uwagura** [reply]
my bad, I meant to include the paths in my original message. v3.1 data is in `/work/Utheri.Wagura/datasets/glofas/v3.1`, and v4 data is in `/work/Utheri.Wagura/datasets/glofas/v4.0`
 
Also, the Local Drain Direction data for v4 is in `/work/Utheri.Wagura/datasets/glofas/LDD`

**Reply by andrew-c-ross** [reply]
I haven't been able to reproduce any substantial problems with the data, though it does seem like xarray takes some time before it starts writing to disk.

```python
import xarray
ds = xarray.open_dataset('/work/Utheri.Wagura/datasets/glofas/v4.0/GloFAS_river_discharge_2023_v4.0.nc')
ds = ds.rename({'valid_time': 'time'})
ds.to_netcdf('/work/acr/tmp/glofas.nc', unlimited_dims='time')
```

```sh
du -sh glofas.nc 
2.3G    glofas.nc

du -sh /work/Utheri.Wagura/datasets/glofas/v4.0/GloFAS_river_discharge_2023_v4.0.nc
2.4G    /work/Utheri.Wagura/datasets/glofas/v4.0/GloFAS_river_discharge_2023_v4.0.nc
```

**Reply by uwagura** [reply]
Thanks for checking this out. It _looks_ like the difference in file sizes was just an artifact of how `du` works on recently created netcdf files. I'm not quite sure what's going on yet, but it looks like if I wait a bit, `du` eventually outputs the correct size of the netcdf file, so this wasn't actually an issue. 

If everything else about these files looks ok, I guess the next steps are just: 

1. Renaming `valid_time` to time
2. Running some of the QA/QC functions I wrote on the GloFAS data, 
3. Moving the data to /uda/GloFas once Kris gives the go ahead. 

Let me know if there is anything else you think I should do!

**Reply by andrew-c-ross** [reply]
Since time isn't an unlimited dimension in the original data, it would be useful (but not essential) if it could be made unlimited while you're renaming valid_time, like I did above with `unlimited_dims='time'` or else `ncks --mk_rec_dmn`. Otherwise sounds great!
